main/lsp/requests/completion.cc

```diff
@@ -18,43 +18,50 @@ namespace {
 struct RubyKeyword {
     const string keyword;
     const string documentation;
+    const optional<string> snippet;
 
-    RubyKeyword(string keyword, string documentation) : keyword(keyword), documentation(documentation){};
+    RubyKeyword(string keyword, string documentation, optional<string> snippet = nullopt)
+        : keyword(keyword), documentation(documentation), snippet(snippet){};
 };
 
 // Taken from https://docs.ruby-lang.org/en/2.6.0/keywords_rdoc.html
 // We might want to put this somewhere shareable if there are more places that want to use it.
+//
+// VS Code snippet syntax is in general smarter than LSP snippet syntax.
+// Specifically, VS Code will intelligently insert the correct indentation after newlines.
 const RubyKeyword rubyKeywords[] = {
     {"BEGIN", "Runs before any other code in the current file."},
     {"END", "Runs after any other code in the current file."},
     {"__ENCODING__", "The script encoding of the current file."},
     {"__FILE__", "The path to the current file."},
     {"__LINE__", "The line number of this keyword in the current file."},
-    {"alias", "Creates an alias between two methods (and other things)."},
+    {"alias", "Creates an alias between two methods (and other things).", "alias ${1:_new} ${2:_existing}$0"},
     {"and", "Short-circuit Boolean and with lower precedence than &&"},
-    {"begin", "Starts an exception handling block."},
+    {"begin", "Starts an exception handling block.", "begin\n  $0\nend"},
     {"break", "Leaves a block early."},
-    {"case", "Starts a case expression."},
-    {"class", "Creates or opens a class."},
-    {"def", "Defines a method."},
-    {"defined?", "Returns a string describing its argument."},
-    {"do", "Starts a block."},
-    {"else", "The unhandled condition in case, if and unless expressions."},
-    {"elsif", "An alternate condition for an if expression."},
+    {"case", "Starts a case expression.", "case ${1:expr}\nwhen ${2:expr}\n  $0\nelse\nend"},
```

Comment:
Does this work as expected?

> Placeholders with equal identifiers are linked, that is typing in one will update others too

Are identifiers equal if the names are the same but numbers are different? (Maybe not.)

Reply:
oh cool TIL

yeah, if the number is different, they're not linked, even if they have the same placeholder text. also, if the number is the same they **are** linked, even if the placeholder text is different.

tl;dr: this snippet works as written in the PR. thanks!